Verify message by sender

rrrooommmaaa opened the pull request:
This PR refactors decryption functions to use verificationPubs as parameter,
fetches unknown sender's pubkeys with publookup, automatically saving them to ContactStore and using to verify signature.
sender email is always displayed as Signer (primary UID from pubkey is ignored)

close #3884

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

tomholub commented on extension/js/common/core/crypto/pgp/msg-util.ts:
```diff
@@ -148,80 +143,33 @@ export class MsgUtil {
       return { success: false, error: { type: DecryptErrTypes.format, message: String(formatErr) }, longids };
     }
+    if (prepared.isCleartext) {
+      // todo: error if no verificationPubs?
```
But what if this is a signed-only pkcs7 message, wouldn't it go into this OpenPGP branch and fail? Mabe need to add another if or another condition in this if so that we handle the issue gracefully.

rrrooommmaaa replied:
`isCleartext` and `isPkcs7` are mutually exclusive, so it won't go to this branch

tomholub replied:
Ah. no such thing as a signed-only pkcs7 message?

rrrooommmaaa replied:
Added appropriate remarks

rrrooommmaaa replied:
Please see the updated comment

tomholub reviewed (commented):
this is roughly it, as a wip

rrrooommmaaa commented:
@tomholub 
Implemented, though there are some differences with the task description

> You'll need to take a method from compose window and put it somewhere more general (it's comparing received pubkeys by fingerprint to the ones saved for that recipient, and only saving the newer versions / new keys, then saving to storage). 

Why don't we want to send all the keys to ContactStore.update, to update `pubkeyLastCheck` property?

> If the fetch fails, we render "error when fetching public keys to verify this message: $error" in the signature result.

It currently says "Could not look up sender's pubkey due to network error, click to retry." or, on `ApiErr.isSignificant()` -- "Could not load sender's pubkey due to an error." Should I change it and write a unit test?

> after remote public keys are fetched and local storage is updated, we check if any of the keys from local storage now contain the longid that signed the message. If they don't, now we render the "missing pub key" signature result

I implemented asynchronous ContactStore update calls -- no need to wait for completion -- is it ok?

> if one of the keys now contain the longid of the key that signed the message, we re-process the whole message with that public key - in background. 

By background do you mean sending PgpMsgVerify? That is implemented.

> Once we've done re-processing the message, we only look at the result of signature verification and re-render the signature result with the final verification result (valid / bad / error), without re-rendering the rest of the message.

Are you concerned about taking too much time to re-render the message? I need to double-check on that. The whole chain of calls DecryptModule->RenderModule->SignatureModule->DecryptModule is far too complicated. Also, I think we SHOULD re-render the message text when re-fetching it from API -- I filed an issue #4164

tomholub commented on extension/chrome/elements/pgp_block_modules/pgp-block-signature-module.ts:
```diff
@@ -3,111 +3,101 @@
+              await this.renderPgpSignatureCheckResult(await retryVerification(fetchedPubkeys), fetchedPubkeys, undefined);
+              return;
+            }
+            if (intersection.length) {
+              this.renderBadSignature();
+            } else {
+              this.renderMissingPubkey(verifyRes.signerLongids[0]);
+            }
```
I don't think you have to look for intersection here. Consider this code from iOS (which is using VerifyRes that comes from TypeScript, originally copied from browser extension, should be very similar code):

```swift
  private func renderPgpSignatureCheckResult(
    signature: VerifyRes?, 
    sender: RecipientWithOrderedPublicKeys
  ) -> ProcessedMessage.MessageSignature {
    if signature is nil then return .unsigned // render "not signed" in red
    if signature.error is not nil then return .error(signature.error) // render the "cannot verify signature: \(error)"
    if signature.signer is nil then return .unsigned // render "not signed" in red
    let signingPubKey = sender.pubkeys.find { $0.longids.contains(signature.signer) }
    if signingPubKey is nil or signature.match is nil then return .missingPubkey(longid: signature.signer) // render "cannot verify signature: no Public Key $longid" in red
    if signature.match is false then return .bad // render "bad signature" in red
    return .good
  }
```

By following the same steps, you should be able to simplify the code and retain correctness.

Please try that, see if that breaks any tests you wrote, in which case let's investigate the delta.

rrrooommmaaa replied:
You're right, I actually only need `hasIntersection` method which is implemented here 
`sender.pubkeys.find { $0.longids.contains(signature.signer) }`

tomholub commented on extension/js/common/browser/browser-msg.ts:
```diff
@@ -148,8 +143,6 @@ export class BrowserMsg {
         pgpHashChallengeAnswer: (bm: Bm.PgpHashChallengeAnswer) => BrowserMsg.sendAwait(undefined, 'pgpHashChallengeAnswer', bm, true) as Promise<Bm.Res.PgpHashChallengeAnswer>,
         pgpMsgDecrypt: (bm: Bm.PgpMsgDecrypt) => BrowserMsg.sendAwait(undefined, 'pgpMsgDecrypt', bm, true) as Promise<Bm.Res.PgpMsgDecrypt>,
         pgpMsgVerifyDetached: (bm: Bm.PgpMsgVerifyDetached) => BrowserMsg.sendAwait(undefined, 'pgpMsgVerifyDetached', bm, true) as Promise<Bm.Res.PgpMsgVerify>,
-        keyParse: (bm: Bm.KeyParse) => BrowserMsg.sendAwait(undefined, 'keyParse', bm, true) as Promise<Bm.Res.KeyParse>,
-        keyMatch: (bm: Bm.KeyMatch) => BrowserMsg.sendAwait(undefined, 'keyMatch', bm, true) as Promise<Bm.Res.KeyMatch>,
```
I like to see stuff removed 👍

tomholub commented on test/source/tests/browser-unit-tests/unit-ContactStore.js:
```diff
@@ -371,57 +299,20 @@ BROWSER_UNIT_TEST_NAME(`ContactStore stores postfixed fingerprint internally for
   return 'pass';
 })();
 
-BROWSER_UNIT_TEST_NAME(`ContactStore searches S/MIME Certificate by PKCS#7 message recipient`);
```
Is this usecase covered elsewhere, or not needed?

rrrooommmaaa replied:
Search by longid is no longer required anywhere in the system. We're now loading all sender's keys and only use them to verify. This should also be the case for S/MIME messages (though verification isn't yet implemented for them)

tomholub replied:
Got it

tomholub commented on test/source/tests/decrypt.ts:
```diff
@@ -371,6 +375,36 @@ export const defineDecryptTests = (testVariant: TestVariant, testWithBrowser: Te
+        await contactsFrame.waitAll('@page-contacts');
+        await Util.sleep(1);
+        expect(await contactsFrame.isElementPresent('@action-show-email-flowcryptcompatibilitygmailcom')).to.be.true;
+        expect(await contactsFrame.isElementPresent('@action-show-email-dhartleyverdoncollegeschoolnz')).to.be.false;
```
Prefer not to have 3rd party or user emails in codebase -better craft our own message for this if needed that has no PII in it.

tomholub commented on test/source/tests/decrypt.ts:
```diff
@@ -371,6 +375,36 @@ export const defineDecryptTests = (testVariant: TestVariant, testWithBrowser: Te
+        expect(await contactsFrame.isElementPresent('@action-show-email-flowcryptcompatibilitygmailcom')).to.be.true;
+        expect(await contactsFrame.isElementPresent('@action-show-email-dhartleyverdoncollegeschoolnz')).to.be.false;
+      }
+      const params = `?frameId=none&acctEmail=${acctEmail}&msgId=${msgId}&signature=___cu_true___&senderEmail=dhartley@verdoncollege.school.nz`;
```
And here.

tomholub commented on test/source/tests/decrypt.ts:
```diff
@@ -394,26 +428,73 @@ export const defineDecryptTests = (testVariant: TestVariant, testWithBrowser: Te
+      await BrowserRecipe.pgpBlockVerifyDecryptedContent(t, browser, {
+        params: url,
+        content: ['How is my message signed?'],
+        signature: ['President@Forged.Com', 'matching signature']
```
Was this the sender of the message in the mock? A bit of an unfortunate choice for a test because it looks like we are testing for forged signatures succeeding, which is not the case :-)

rrrooommmaaa replied:
Renamed

tomholub commented on test/source/tests/decrypt.ts:
```diff
@@ -394,26 +428,73 @@ export const defineDecryptTests = (testVariant: TestVariant, testWithBrowser: Te
+      await BrowserRecipe.pgpBlockVerifyDecryptedContent(t, browser, {
+        params: url,
+        content: ['How is my message signed?'],
+        signature: ['President@Forged.Com', 'matching signature']
```
Also, we should remove the capitalization from the email address rendering in signature result. It was a bad idea.

tomholub commented on test/source/tests/decrypt.ts:
```diff
@@ -394,26 +428,73 @@ export const defineDecryptTests = (testVariant: TestVariant, testWithBrowser: Te
+    ava.default('decrypt - re-fetch signed-only message from API on non-fatal verification error', testWithBrowser('compatibility', async (t, browser) => {
+      const msgId = '1754cfd1b2f1d6e5';
+      const acctEmail = 'flowcrypt.compatibility@gmail.com';
+      const signerEmail = 'dhartley@verdoncollege.school.nz';
```
Here PII better not include

tomholub commented on test/source/tests/decrypt.ts:
```diff
@@ -394,26 +428,73 @@ export const defineDecryptTests = (testVariant: TestVariant, testWithBrowser: Te
+      await BrowserRecipe.pgpBlockVerifyDecryptedContent(t, browser, {
+        params,
+        content: [], // todo: #4164 I would expect '1234' here
+        signature: ['Dhartley@Verdoncollege.School.Nz', 'matching signature']
```
And here

tomholub commented on test/source/tests/decrypt.ts:
```diff
@@ -394,26 +428,73 @@ export const defineDecryptTests = (testVariant: TestVariant, testWithBrowser: Te
+      // as the verification pubkey is retrieved from the attester, the incorrect message digest will trigger re-fetching from API
+      await BrowserRecipe.pgpBlockVerifyDecryptedContent(t, browser, {
+        params,
+        content: [], // todo: #4164 I would expect '1234' here
```
Highlighting todo. We'll need a new message anyway

rrrooommmaaa replied:
I constructed a new message containing '1234'. Here my concern is whether the "slightly modified" originally loaded message should be updated (in the end we would like to show signature next to the really signed message, don't we?)

tomholub replied:
Let's discuss this in main comments below

tomholub commented on test/source/tests/unit-node.ts:
```diff
@@ -776,19 +775,40 @@ jLwe8W9IMt765T5x5oux9MmPDXF05xHfm4qfH/BMO3a802x5u2gJjJjuknrFdgXY
-      expect(verifyRes.match).to.be.true;
-      expect(verifyRes.signer?.primaryUserId).to.equal('A50 Sam <sams50sams50sept@gmail.com>');
+      // actual key the message was signed with
+      const pubkey = testConstants.pubkey2864E326A5BE488A;
+      // better key
+      const betterKey = "-----BEGIN PGP PUBLIC KEY BLOCK-----\r\nVersion: FlowCrypt Email Encryption [BUILD_REPLACEABLE_VERSION]\r\nComment: Seamlessly send and receive encrypted email\r\n\r\nxjMEYZeW2RYJKwYBBAHaRw8BAQdAT5QfLVP3y1yukk3MM/oiuXLNe1f9az5M\r\nBnOlKdF0nKnNJVNvbWVib2R5IDxTYW1zNTBzYW1zNTBzZXB0QEdtYWlsLkNv\r\nbT7CjwQQFgoAIAUCYZeW2QYLCQcIAwIEFQgKAgQWAgEAAhkBAhsDAh4BACEJ\r\nEMrSTYqLk6SUFiE ...
```
What is the difference between the keys, how is the other one better? Would be good to explain in comments.

rrrooommmaaa replied:
The newest (the greater date of creation) was supposed to be "better", but currently this field isn't used in sorting. Should we take it into account?

tomholub replied:
Let's name them `olderVersionOfPubkey` and `newerVersionOfPubkey` or similar. Better is unclear

rrrooommmaaa replied:
Done. Although they are not really versions -- they are compoletely independent. Is `newlyCreatedPubkey` good enough?

tomholub replied:
> The newer (the greater date of creation) was supposed to be "better", but currently this field isn't used in sorting. Should we take it into account?

Newer key is only better if both have the exact same primary fingerprint, and therefore the newer is an updated version of the old. There is nothing better about a newly created different key. It's just another key.\

Do we have anywhere in our codebase a preference for newer keys that have a different primary fingerprint? If yes, that has to be removed (sounds like a security problem).

For this test, let's call the second key simply `wrongKey`. Since it's a key unrelated to the signature that was made. Whether it's newer or not should not matter.

rrrooommmaaa replied:
> Do we have anywhere in our codebase a preference for newer keys that have a different primary fingerprint? If yes, that has to investigated and discussed.

No, it turns out we don't have such preference.
We sort by expiration date. Both keys in this test are non-expirable, so they are considered equivalent

tomholub replied:
Got it. Let's use `wrongKey` as a name for the second one.

rrrooommmaaa replied:
Done

tomholub commented on test/source/tests/unit-node.ts:
```diff
@@ -844,15 +862,52 @@ jLwe8W9IMt765T5x5oux9MmPDXF05xHfm4qfH/BMO3a802x5u2gJjJjuknrFdgXY
+      const msg = data.getMessage('1754cfc37886899e')!;
+      const msgText = Buf.fromBase64Str(msg!.raw!).toUtfStr();
+      {
+        const dhartleyPubkey = msgText
```
And here PII

tomholub commented on test/source/tests/unit-node.ts:
```diff
@@ -844,15 +862,52 @@ jLwe8W9IMt765T5x5oux9MmPDXF05xHfm4qfH/BMO3a802x5u2gJjJjuknrFdgXY
+          verificationPubs: [dhartleyPubkey]
         });
-        await ContactStore.save(undefined, contact);
+        expect(resultRightKey.match).to.be.null;
```
Originally I imagine this would be a `false` ? Have the right pubkey, but didn't verify succcesfully.

rrrooommmaaa replied:
Fixed

tomholub commented:
Thank you for the PR - I hope to do some manual testing tomorrow 👍

rrrooommmaaa commented on extension/chrome/elements/pgp_block_modules/pgp-block-signature-module.ts:
```diff
@@ -3,111 +3,101 @@
+              }
+              await this.renderPgpSignatureCheckResult(await retryVerification(fetchedPubkeys), fetchedPubkeys, undefined);
+              return;
+            }
```
Done. I exported functions directly, should I have used a class with static members instead?

rrrooommmaaa commented on extension/chrome/elements/pgp_block_modules/pgp-block-decrypt-module.ts:
```diff
@@ -54,39 +54,37 @@ export class PgpBlockViewDecryptModule {
-    if (typeof this.view.signature !== 'string') {
+  public canFetchFromApi = () => this.canReadEmails && this.msgFetchedFromApi !== 'raw';
+
+  private decryptAndRender = async (encryptedData: Buf, verificationPubs: string[], optionalPwd?: string, plainSubject?: string) => {
```
fixed

tomholub reviewed (commented):
New changes all look good. I'll do another review for all of the changes hollistically

tomholub commented:
> This pull request **introduces 1 alert** when merging [2df3953](https://github.com/FlowCrypt/flowcrypt-browser/commit/2df39531610a2cd28a31d4dd1cb49b6e101be2ce) into [c636c10](https://github.com/FlowCrypt/flowcrypt-browser/commit/c636c103cfbef2171c3a22497f0fca6c1b052838) - [view on LGTM.com](https://lgtm.com/projects/g/FlowCrypt/flowcrypt-browser/rev/pr-120585728db0347a98592a13974cc9f22f9a5cde)
> 
> **new alerts:**
> 
>     * 1 for User-controlled bypass of security check

@rrrooommmaaa have you looked into this - is that a false positive?

rrrooommmaaa commented:
> have you looked into this - is that a false positive?

Looks like this means that `verify` call is skipped if the module is invoked without `signature` query parameter (I guess it was the same way before), if I understood correctly. Anyway, if `verify` isn't called, then `matching signature` status won't appear, so I can't see any problems here.

tomholub commented:
![image](https://user-images.githubusercontent.com/6306961/146226775-8adba033-93fc-4f0b-ad5b-f337ffefd69b.png)

If you are talking about re-fetching the message because the DOM may have mingled it, then yes - we should display the re-fetched message we got from Google API.

If you mean re-fetching public keys for verification, then we don't have to re-render the message, just verification result.

Does this mean this is still actionable to change / update in this PR?

rrrooommmaaa commented:
> If you are talking about re-fetching the message because the DOM may have mingled it, then yes - we should display the re-fetched message we got from Google API.

yes, in this particular test the message is completely different when re-fetched from API.

tomholub commented:
> > If you are talking about re-fetching the message because the DOM may have mingled it, then yes - we should display the re-fetched message we got from Google API.
> 
> yes, in this particular test the message is completely different when re-fetched from API.

We should indeed re-render the message body to match what we got from API. Since that's what we are verifying on the second attempt. Could be in another issue - since we were never doing that before, it's not new to this PR. Well noticed.

rrrooommmaaa commented:
> If you mean re-fetching public keys for verification, then we don't have to re-render the message, just verification result.

Is this for performance reasons. I didn't pay too much attention on what is rendered and re-rendered. I can do this in this or another PR

rrrooommmaaa commented:
> We should indeed re-render the message body to match what we got from API. Since that's what we are verifying on the second attempt. Could be in another issue - since we were never doing that before, it's not new to this PR. Well noticed.

The issue is already filed (#4164)

tomholub commented:
> > If you mean re-fetching public keys for verification, then we don't have to re-render the message, just verification result.
> 
> Is this for performance reasons. I didn't pay too much attention on what is rendered and re-rendered. I can do this in this or another PR

It's just for less work reasons, development-wise. Usually re-rendering fewer things is less work. As long as it doesn't create any side effects, it doesn't really matter if the body is re-rendered.

There is one possible side effect - if the user starts wanting to copy and paste part of the message, re-rendering will upset them because their highlight will be gone if the text gets replaced while trying to copy it.

rrrooommmaaa commented:
> There is one possible side effect - if the user starts wanting to copy and paste part of the message, re-rendering will upset them because their highlight will be gone if the text gets replaced while trying to copy it.

I'm afraid there might be such side effects. I suggest to fix them after (or as part of) #4164 as there is a stack of modules calling each other. Perhaps, a bit of refactoring is in order.

tomholub commented:
Getting an exception - this scenario is likely not tested? it's on `flowcrypt.compatibility` account.

```
TypeError: Cannot set properties of undefined (setting 'parsedSignature')
    at PgpBlockViewSignatureModule.renderPgpSignatureCheckResult (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block_modules/pgp-block-signature-module.js:17:57)
    at PgpBlockViewRenderModule.decideDecryptedContentFormattingAndRender (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block_modules/pgp-block-render-module.js:70:45)
    at PgpBlockViewDecryptModule.decryptAndRender (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block_modules/pgp-block-decrypt-module.js:69:50)
    at async PgpBlockViewDecryptModule.initialize (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block_modules/pgp-block-decrypt-module.js:34:21)
    at async PgpBlockView.render (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block.js:44:17)
    at async chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/js/common/view.js:26:17

### Catch.reportErr calling stack ###
#     at Function.Catch.formatExceptionForReport (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/js/common/platform/catch.js:252:90)
#     at Function.Catch.onErrorInternalHandler (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/js/common/platform/catch.js:76:29)
#     at Function.Catch.reportErr (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/js/common/platform/catch.js:92:18)
#     at PgpBlockViewErrorModule.handleInitializeErr (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block_modules/pgp-block-error-module.js:56:23)
#     at PgpBlockViewDecryptModule.initialize (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block_modules/pgp-block-decrypt-module.js:55:45)
#     at async PgpBlockView.render (chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block.js:44:17)
#     at async chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/js/common/view.js:26:17
# 
# url: chrome-extension://pljjhfjdhkbkhhkmngllhdmccdflgnbp/chrome/elements/pgp_block.htm?frameId=frame_tXWErApBbb&message=-----BEGIN%20PGP%20SIGNED%20MESSAGE-----%0AHash%3A%20SHA256%0A%0Athis%20message%20is%20signed%20and%20was%20modified%20in%20transit%0A-----BEGIN%20PGP%20SIGNATURE-----%0AVersion%3A%20FlowCrypt%205.0.4%20Gmail%20Encryption%20flowcrypt.com%0AComment%3A%20Seamlessly%20send%2C%20receive%20and%20search%20encrypted%20email%0A%0AwsFcBAEBCAAQBQJZ%2B8VXCRAGylU%2BwkVdcAAAgRwQAMkc%2B2WrIBf%2BX3T%2ByNps%0Ava4xMQEVvBFS4GxbTA1xndmfuo3hr3Vfq%2FsAI2Plo65D0rd9Wc6p88Awl%2B9w%0Aa0l4Yx3isH0jm9gcpypzYJf9leertBcvIxDvqGZc%2FxPxyRzQKgXiNkvaZI8U%0ANNkoAdCves4wx3sb0lJhIKyRhXwKWWroWKfoBYAryKDSE2q8d1z%2BP0mDbop7%0AJ5YLui%2B6VsGJoqZdM8lMGLp2TuyGVxy5moEvrIY8zvYm61Ke%2BwGLbL63Bqdg%0AZb8pMEp1xdW%2FxPgFT6IrtFdHTEiqczkeiHOG7%2BF8zPcKBbvx6luFAvFh103I%0Al06RRefeJkppLJ2S%2BA0iDzmqDknYin2Hm8VBa2ill%2FRfI4c1ThzM5trjrC1b%0ArOMzIMZfDdCDCRjfILNyjnanhsYEYZ30NnmjOrNDthA9VCEUupoRvSmDbA8O%0AVtckZH3YzIi6CsHs8dl2WTQ%2Flj1yXKDqnpoWzndzVrFRtjVW7I75lL%2F%2B%2FMYp%0AhktbFAWIkrqqWapwIUEGefzbrulRRVUU6m6slbKIurOgkEt%2BcoJUNMTsuPPX%0AaAa43ze8hUji5EXALE9I%2B41uz%2BdtnJymaQhsnXL8R8VHAgowrvQgLeuTZNVH%0AAeNkbcng%2BKz3ErVk3Q%2BRDt0i5FL0oIzP%2FAsxhGeOG38j%2BQVe6OgYm7IjfaIA%0Ar498%0A%3D76q6%0A-----END%20PGP%20SIGNATURE-----&msgId=15f7f7c5979b5a26&senderEmail=cryptup.tester%40gmail.com&isOutgoing=___cu_false___&acctEmail=flowcrypt.compatibility%40gmail.com&parentTabId=73%3A0
# 
######################
```

![image](https://user-images.githubusercontent.com/6306961/146229292-555253d6-23a9-44b7-a637-41851ea98825.png)

tomholub commented:
Also happens on `[enigmail] signed message inline + attachment`

tomholub commented:
Separately, it seems to me that the pgp block doesn't finish rendering anything until the decrypt -> try to verify -> aha missing pubkey -> fetch pubkeys -> reverify loop is complete.

I didn't investigate in detail, could it be true? I'd recommend to write a test as follows:
 - add `/attester/pub/this.pubkey.takes.long.to.load@sender.test` to attester mock api. It will include a 5 second sleep before returning the pubkey.
 - write a ui test that checks a message signed by that key
 - wait in ui test of the pgp block for texts in this sequence: 1) message body 2) "verifying signature" 3) good signature

Do you have a similar test that checks for the "verifying signature" step specifically?

tomholub commented:
> >  if one of the keys now contain the longid of the key that signed the message, we re-process the whole message with that public key - in background.
> By background do you mean sending PgpMsgVerify? That is implemented.

By background I meant that the user should not need to wait for it. Render the text first with "verifying signature..". So user can see message right away, and signature verification when it's done verifying.

tomholub commented:
> >  Once we've done re-processing the message, we only look at the result of signature verification and re-render the signature result with the final verification result (valid / bad / error), without re-rendering the rest of the message.
>
> Are you concerned about taking too much time to re-render the message? I need to double-check on that. The whole chain of calls DecryptModule->RenderModule->SignatureModule->DecryptModule is far too complicated. Also, I think we SHOULD re-render the message text when re-fetching it from API -- I filed an issue #4164

Here I was concerned that the user would have to wait 10 seconds to read the message if it takes 10 seconds to fetch public keys to verify it. Sorry for not clarifying earlier, I overlooked this comment.

rrrooommmaaa commented:
> Getting an exception - this scenario is likely not tested? it's on `flowcrypt.compatibility` account.

fixed this -- no longer re-fetching the encrypted message when verification fails

rrrooommmaaa commented:
> Also happens on `[enigmail] signed message inline + attachment`

This one should now be passing too

tomholub commented:
This does not close the issue fully yet - I've filed #4197 which should be done right after.
